### DISCUSSION TITLE
Lastfm - Missing arguments in function application: [string * string].

### DISCUSSION THREAD

**Discussion by oranjebloom** [savonet/liquidsoap#2543] [User Support]
I've upgraded from version 2.0.4 to 2.1.0 via opam on Debian 10.

Since doing so, I get the error (concerning lastfm) below and my script won't start.

```
radio.on_metadata(lastfm.submit(user="****",password="*****"))
Error 15: Missing arguments in function application: [string * string].
```
The previously working section of my script is provided below. Is this a bug or are edits required in order to make this work in version 2.1.0?

```
def lastfm.submit(~user,~password,~source="broadcast",~length=false,m) =
  audioscrobbler.submit(user=user,password=password,
                source=source,length=length,
                host="post.audioscrobbler.com",port=80,
                m)
end

radio.on_metadata(lastfm.submit(user="****",password="*****"))
```

**Comment by vitoyucepi** [comment]
I assume this happened because of removal of partial application in release 2.1.0.

```ruby
radio.on_metadata(fun(m) -> lastfm.submit(user="****", password="*****", m))
```

In 2.0 `lastfm.submit(user='', password='')` creates a new function `t(~source="broadcast", ~length=false, m)`, so `radio.on_metadata` will call `t` with argument `m:[source * source]`.

In 2.1 you have to write a function with all arguments at once.

As for partial application imagine function `f(a, b, c){return a+b+c}`.
You could call it as `f(1,2,3)`, as `f(1,2)(3)`, `f(1)(2)(3)`, etc.
`f(1,2,3)` will instantly return you an answer, but if you pass only some arguments, then `f` will return another function. `f(1,2)` will produce `t(c){return 1+2+c}`.

**Comment by toots** [comment]
@vitoyucepi is correct. Hopefully this shouldn't be too difficult to fix:
```ruby
radio.on_metadata(fun (m) -> lastfm.submit(user="****",password="*****", m))
```

**Comment by toots** [comment]
Alternatively:
```ruby
def lastfm.submit(~user,~password,~source="broadcast",~length=false) = fun (m) ->
  audioscrobbler.submit(user=user,password=password,
                source=source,length=length,
                host="post.audioscrobbler.com",port=80,
                m)
end

radio.on_metadata(lastfm.submit(user="****",password="*****"))
```

**Reply by oranjebloom** [reply]
Thank you @toots and @vitoyucepi that sorted out the problem.
